common/cmake/devices.cmake

```diff
@@ -929,6 +929,186 @@ function(ADD_FPGA_TARGET_BOARDS)
+    get_target_property_required(PACKAGE ${OVERLAY_BOARD} PACKAGE)
+    get_target_property_required(PYTHON3 env PYTHON3)
+
+    get_target_property(FASM_TO_BIT_EXTRA_ARGS ${OVERLAY_BOARD} FASM_TO_BIT_EXTRA_ARGS)
+    if ("${FASM_TO_BIT_EXTRA_ARGS}" STREQUAL "FASM_TO_BIT_EXTRA_ARGS-NOTFOUND")
+      set(FASM_TO_BIT_EXTRA_ARGS "")
+    endif()
+
+    get_target_property_required(TOP ${OVERLAY} TOP)
+    get_target_property_required(ARCH ${OVERLAY_DEVICE} ARCH)
+    get_target_property_required(BITSTREAM_EXTENSION ${ARCH} BITSTREAM_EXTENSION)
+    get_target_property_required(FASM_TO_BIT ${ARCH} FASM_TO_BIT)
+    get_target_property_required(FASM_TO_BIT_CMD ${ARCH} FASM_TO_BIT_CMD)
+    get_target_property(FASM_TO_BIT_DEPS ${ARCH} FASM_TO_BIT_DEPS)
+
+    if(NOT USE_OLD_DIR)
+      set(FQDN ${ARCH}-${DEVICE_TYPE}-${DEVICE}-${PACKAGE})
+      set(OUT_LOCAL_REL ${NAME}/${FQDN})
+      set(OUT_LOCAL ${CMAKE_CURRENT_BINARY_DIR}/${OUT_LOCAL_REL})
+      set(OVERLAY ${NAME})
+      add_custom_target(${NAME})
+    else()
+      set(OUT_LOCAL ${USE_OLD_DIR})
+    endif()
+
+    set(OUT_FASM_MERGED ${OUT_LOCAL}/${TOP}.merged.fasm)
+    set(OUT_BITSTREAM ${OUT_LOCAL}/${TOP}.${BITSTREAM_EXTENSION})
+
+    set(BITSTREAM_DEPS ${ALL_OUT_FASM_DEPS} ${FASM_TO_BIT} ${FASM_TO_BIT_DEPS})
+
+    set(OUT_FASM ${OUT_FASM_MERGED})
+    string(CONFIGURE ${FASM_TO_BIT_CMD} FASM_TO_BIT_CMD_FOR_TARGET)
+    separate_arguments(
+      FASM_TO_BIT_CMD_FOR_TARGET_LIST UNIX_COMMAND ${FASM_TO_BIT_CMD_FOR_TARGET}
+    )
+
+    separate_arguments(
+      FASM_TO_BIT_EXTRA_ARGS_LIST UNIX_COMMAND ${FASM_TO_BIT_EXTRA_ARGS}
+    )
+    set(FASM_TO_BIT_CMD_FOR_TARGET_LIST ${FASM_TO_BIT_CMD_FOR_TARGET_LIST} ${FASM_TO_BIT_EXTRA_ARGS_LIST})
+
+    if(NOT USE_OLD_DIR)
+      add_custom_command(
+        OUTPUT ${OUT_BITSTREAM}
+        DEPENDS ${BITSTREAM_DEPS}
+        COMMAND
+          ${CMAKE_COMMAND} -E make_directory ${OUT_LOCAL}
+        COMMAND cat ${ALL_OUT_FASM} >> ${OUT_FASM_MERGED}
+        COMMAND ${FASM_TO_BIT_CMD_FOR_TARGET_LIST}
+      )
+    else()
+      add_custom_command(
+        OUTPUT ${OUT_BITSTREAM}
+        DEPENDS ${BITSTREAM_DEPS}
+        COMMAND cat ${ALL_OUT_FASM} >> ${OUT_FASM_MERGED}
+        COMMAND ${FASM_TO_BIT_CMD_FOR_TARGET_LIST}
+      )
+    endif()
+  else()
+    get_target_property_required(HLC_TO_BIT ${ARCH} HLC_TO_BIT)
+    get_target_property_required(HLC_TO_BIT_CMD ${ARCH} HLC_TO_BIT_CMD)
+    string(CONFIGURE ${HLC_TO_BIT_CMD} HLC_TO_BIT_CMD_FOR_TARGET)
+    separate_arguments(
+      HLC_TO_BIT_CMD_FOR_TARGET_LIST UNIX_COMMAND ${HLC_TO_BIT_CMD_FOR_TARGET}
+    )
+    add_custom_command(
+      OUTPUT ${OUT_BITSTREAM}
+      DEPENDS ${OUT_HLC} ${HLC_TO_BIT}
+      COMMAND ${HLC_TO_BIT_CMD_FOR_TARGET_LIST}
+    )
+  endif()
+
+  add_custom_target(${OVERLAY}_bit DEPENDS ${OUT_BITSTREAM})
+  add_output_to_fpga_target(${OVERLAY} BIT ${OUT_LOCAL}/${TOP}.${BITSTREAM_EXTENSION})
```

Comment:
This should be `OUT_LOCAL_REL`

Reply:
The use of `${OVERLAY}` here is confusing.  `${OVERLAY}` should be something like `OUTPUT_TARGET`.  What happens with the current code if you want to generate two bitstreams with the same overlay, wouldn't this cause a conflict?

Reply:
Feel like this should be `${OVERLAY}` should be `${NAME}` here, with the assumption that the target already exists?  Or maybe controlled by an option?

Reply:
> Feel like this should be `${OVERLAY}` should be `${NAME}` here, with the assumption that the target already exists? Or maybe controlled by an option?

My intention is that ${OVERLAY} should either be the target of the overlay architecture if multiple architecture are included, or just the target given if only one architecture is provided.  I don't think this works quite right now so I am working on a fix.

>The use of ${OVERLAY} here is confusing. ${OVERLAY} should be something like OUTPUT_TARGET.
>
>Additional concern, what happens with the current code if you want to generate two bitstreams with the same overlay, wouldn't this cause a conflict?

I agree OVERLAY is confusing.  Will fix.  I think it should be fine to generate two bitstreams with the same overlay because you can provide a different name and have it placed in a different folder.